src/ddmd/astbase.d

```diff
@@ -0,0 +1,6303 @@
+module ddmd.astbase;
+
+import ddmd.astbasevisitor;
+
+/** The ASTBase  family defines a family of AST nodes appropriate for parsing with
+  * no semantic information. It defines all the AST nodes that the parser needs
+  * and also all the conveniance methods and variables. The resulting AST can be
+  * visited with the strict, permissive and transitive visitors.
+  * The ASTBase family is used to instantiate the parser in the parser library.
+  */
+struct ASTBase
+{
+    import ddmd.root.file;
+    import ddmd.root.filename;
+    import ddmd.root.array;
+    import ddmd.root.rootobject;
+    import ddmd.root.outbuffer;
+    import ddmd.root.ctfloat;
+    import ddmd.root.rmem;
+    import ddmd.root.stringtable;
+
+    import ddmd.tokens;
+    import ddmd.identifier;
+    import ddmd.globals;
+    import ddmd.id;
+    import ddmd.errors;
+    import ddmd.lexer;
+
+    import core.stdc.string;
+    import core.stdc.stdarg;
+
+    alias Dsymbols              = Array!(Dsymbol);
+    alias Objects               = Array!(RootObject);
+    alias Expressions           = Array!(Expression);
+    alias TemplateParameters    = Array!(TemplateParameter);
+    alias BaseClasses           = Array!(BaseClass*);
+    alias Parameters            = Array!(Parameter);
+    alias Statements            = Array!(Statement);
+    alias Catches               = Array!(Catch);
+    alias Identifiers           = Array!(Identifier);
+    alias Initializers          = Array!(Initializer);
+
+    enum PROTKIND : int
+    {
+        PROTundefined,
+        PROTnone,
+        PROTprivate,
+        PROTpackage,
+        PROTprotected,
+        PROTpublic,
+        PROTexport,
+    }
+
+    alias PROTprivate       = PROTKIND.PROTprivate;
+    alias PROTpackage       = PROTKIND.PROTpackage;
+    alias PROTprotected     = PROTKIND.PROTprotected;
+    alias PROTpublic        = PROTKIND.PROTpublic;
+    alias PROTexport        = PROTKIND.PROTexport;
+    alias PROTundefined     = PROTKIND.PROTundefined;
+    alias PROTnone          = PROTKIND.PROTnone;
+
+    enum Sizeok : int
+    {
+        SIZEOKnone,             // size of aggregate is not yet able to compute
```

Comment:
Ideally all symbols should be documented but perhaps we can start by turning the existing comments, like these, to ddoc comments.